src/merkle_tree.rs

```diff
@@ -30,142 +47,131 @@ impl MerkleNode {
     }
 }
 
-impl<H: Hash + Clone> MerkleTree<H> {
+impl LeafNode {
+    pub fn new(hash_value: String) -> Self {
+        Self { hash_value }
+    }
+}
+
+impl<H: AsRef<[u8]> + Clone> MerkleTree<H> {
     pub fn new(transactions: Vec<H>) -> Result<Self, &'static str> {
         Self::create_tree(transactions)
     }
 
     // Fathers must have at least one son, if it does not have one, we clone the left one
     fn create_parent_from_siblings(
-        left_son: Box<MerkleNode>,
-        mut right_son: Option<Box<MerkleNode>>,
+        left_son: MerkleNode,
+        right_son: Option<MerkleNode>,
     ) -> MerkleNode {
-        let mut hasher = DefaultHasher::new();
-        left_son.hash_value.hash(&mut hasher);
-
-        match &right_son {
-            Some(right_sibling) => {
-                right_sibling.hash_value.hash(&mut hasher);
-            }
-            None => {
-                right_son = Some(left_son.clone());
-                left_son.hash_value.hash(&mut hasher);
-            }
-        }
-
-        MerkleNode::new(hasher.finish(), Some(left_son), right_son)
+        let mut hasher = Sha3_256::new();
+        hasher.update(left_son.get_hash_value());
+        let right_son = right_son.unwrap_or_else(|| left_son.clone());
+        hasher.update(right_son.get_hash_value());
+
+        MerkleNode::Inner(InnerNode::new(
+            format!("{:X}", hasher.finalize()),
+            Rc::new(left_son),
+            Rc::new(right_son),
+        ))
     }
 
     fn create_tree(transactions: Vec<H>) -> Result<MerkleTree<H>, &'static str> {
         if transactions.is_empty() {
             return Err("Can't create a tree without elements");
         }
 
-        let transactions_hash = Self::get_hashes_of_transactions(&transactions);
-        let mut nodes: Vec<Box<MerkleNode>> = transactions_hash
+        let mut nodes: Vec<MerkleNode> = transactions
+            .clone()
             .into_iter()
-            .map(|hash| Box::new(MerkleNode::new(hash, None, None)))
+            .map(|transaction| {
+                let mut hasher = Sha3_256::new();
+                hasher.update(transaction);
+                MerkleNode::Leaf(LeafNode::new(format!("{:X}", hasher.finalize())))
+            })
```

Comment:
There is no need for `transactions.clone()` if you're only iterating it once and `collect()`ing the results. Use `transactions.iter()` instead.

Reply:
Solved!